File: flunky.js
const randFloat = (min, max) => Math.random() * (max - min) + min;
const randInt = (min, max) => Math.floor(randFloat(min, max));
const choice = (array) => array[randInt(0, array.length)];
const choices = function (array, amt, joined = false, delim = '') {
  let temp = [];
  for (let i = 0; i < amt; i++) {
    temp.push(choice(array));
  }
  return joined ? temp.join(delim) : temp;
};

const DiceFlags = {
  None: 0,
  TakeHighest: 1,
  DropHighest: 2,
  TakeLowest: 3,
  DropLowest: 4,
};

class RNG {
  constructor(min, max) {
    this.min = min;
    this.max = max;
  }

  getRand(n = 1) {
    let temp = [];
    for (let i = 0; i < n; i++) {
      temp.push(randInt(this.min, this.max));
    }
    return temp;
  }
}

const sum = (arr) => arr.reduce((a, b) => a + b);
const dropLowest = (arr) => arr.filter((x) => x != Math.min(arr));
const dropHighest = (arr) => arr.filter((x) => x != Math.max(arr));
const joinString = (arr) => arr.reduce((a, b) => a + b);

function get(arr, idx) {
  if (idx < 0) {
    return arr[arr.length + idx];
  }
  return arr[idx];
}

class DiceFlagHandler {
  constructor(arr) {
    this.arr = arr;
    this.conversionTable = {
      'd': DiceFlags.None,
      'd!': DiceFlags.TakeHighest,
      'd*': DiceFlags.DropHighest,
      'd_': DiceFlags.TakeLowest,
      'd@': DiceFlags.DropLowest,
    };
  }

  convert(flagStr) {
    return this.conversionTable[flagStr];
  }

  getFromFlag(flag) {
    const A = this.arr;
    switch (flag) {
      case DiceFlags.None:
        return sum(A);
      case DiceFlags.TakeHighest:
        return Math.max(...A);
      case DiceFlags.TakeLowest:
        return Math.min(...A);
      case DiceFlags.DropHighest:
        return dropHighest(A);
      case DiceFlags.DropLowest:
        return dropLowest(A);
      default:
        return sum(A);
    }
  }
}

class DiceParser {
  constructor(dice) {
    this.dice = dice;
  }

  parse() {
    let t = [];
    const m = Array.from(this.dice.matchAll(/(\d+)(d)([^\d]+)?(\d+)/gm));
    m.forEach((e) => {
      t.push([...e.slice(1), e[0]]);
    });
    this.m = t;
    return t;
  }

  format() {
    let curEq = this.dice;
    const r = this.m.map((e) => e.slice(-1));
    for (let i = 0; i < r.length; i++) {
      curEq = curEq.replace(r[i], `$${i}`);
    }
    return curEq.replace(';', '');
  }
}

class DiceRoller {
  constructor(dice) {
    this.dice = dice;
    let x = new DiceParser(dice);
    this.m = x.parse();
    this.eq = x.format();
  }

  roll(e = true) {
    let n = this.eq;
    let i = 0;
    for (const d of this.m) {
      const rand = new RNG(1, parseInt(get(d, -2)));
      const nums = rand.getRand(parseInt(d[0]));
      const flagHandler = new DiceFlagHandler(nums);
      const flagStr = joinString(d.slice(1, -2));
      const flag = flagHandler.convert(flagStr);
      const num = flagHandler.getFromFlag(flag);
      n = n.replace(`$${i}`, num.toString());
      i++;
    }

    return e ? eval(n) : n;
  }
}

class Dice {
  constructor(dice) {
    this.dice = dice;
    this.roller = new DiceRoller(dice);
  }

  roll(e = true) {
    return this.roller.roll((e = e));
  }

  rollN(n, e = true) {
    let t = [];
    for (let i = 0; i < n; i++) {
      t.push(this.roll((e = e)));
    }
    return t;
  }
}

module.exports = { Dice };

File: flunky.py
import secrets
import re
from enum import Enum

class RNG:
  def __init__(self, min, max):
    self.min = int(min)
    self.max = int(max)
    self.rand = secrets.SystemRandom()

  def getRand(self, n=1):
    return [self.rand.randrange(self.min, self.max+1) for _ in range(int(n))]


class DiceFlags(Enum):
  NONE = 0
  TAKE_HIGHEST = 1
  TAKE_LOWEST = 2
  DROP_HIGHEST = 3
  DROP_LOWEST = 4

class DiceFlagHandler:
  def __init__(self, arr):
    self.arr = arr
    self.conversionTable = {
      'd': DiceFlags.NONE,
      'd!': DiceFlags.TAKE_HIGHEST,
      'd*': DiceFlags.DROP_HIGHEST,
      'd_': DiceFlags.TAKE_LOWEST,
      'd@': DiceFlags.DROP_LOWEST
    }

  def convert(self, flagStr):
    return self.conversionTable[flagStr]

  def get(self, flag):
    A = self.arr

    if flag == DiceFlags.TAKE_HIGHEST:
      return max(A)
    elif flag == DiceFlags.TAKE_LOWEST:
      return min(A)
    elif flag == DiceFlags.DROP_HIGHEST:
      return sum(x if x != max(A) else 0 for x in A)
    elif flag == DiceFlags.DROP_LOWEST:
      return sum(x if x != min(A) else 0 for x in A)

    return sum(A)

class DiceParser:
  def __init__(self, dice):
    self.dice = dice

  def parse(self):
    return re.findall(r"(\d+)(d)([^\d]+)?(\d+)", self.dice)

class DiceFormatter:
  def __init__(self, dice):
    self.dice = dice

  def format(self):
    n = 0
    curEq = self.dice
    m = re.findall("\d+d(?:[^\d]+)?\d+", self.dice)
    for match in m:
      curEq = curEq.replace(match, f'${n}')
      n += 1

    return curEq.replace(';', '')

class DiceRoller:
  def __init__(self, dice):
    self.dice = dice
    f = DiceFormatter(dice)
    self.eq = f.format()

  def roll(self, e=True):
    diceParser = DiceParser(self.dice)
    n = self.eq
    for i, d in enumerate(diceParser.parse()):
      rand = RNG(1, d[-1])
      nums = rand.getRand(n=d[0])
      flagHandler = DiceFlagHandler(nums)
      flag = flagHandler.convert(''.join(d[1:-1]))
      num = flagHandler.get(flag)

      n = n.replace(f'${i}', str(num))

    return eval(n) if e else n

class Dice:
  def __init__(self, dice):
    self.dice = dice
    self.roller = DiceRoller(dice)

  def roll(self, e=True):
    return self.roller.roll(e=e)

  def rollN(self, n, e=True):
    t = []
    for _ in range(n):
      t.append(self.roll(e=e))

    return t
